### PR TITLE
Implemented opcodes MSTORE, MSTORE8, MLOAD, MSIZE

### PR DESCRIPTION
**Pull request by voith**
### What was wrong?

Related to Issue #228
- Memory-related operations of the EVM weren't implemented. 

### How was it fixed?
- Implemented memory expansion calculator
- Implemented opcodes `MSTORE`, `MSTORE8`, `MLOAD`, `MSIZE`

## TODO
- [x] Check if more tests can be added.  
- [x] See if some validations need to be added before writing and reading from memory.
- [x] Move the helpers used in the tests to a common file so that they can be re-used.

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: whitelist.txt
Comment (SamWilsn) [inline whitelist.txt:222]: I don't see where this is used.

**Review by SamWilsn — COMMENTED** [review]
Looking good so far!

File: src/ethereum/frontier/vm/instructions.py
Comment (voith) [reply src/ethereum/frontier/vm/instructions.py]: This is to make sure that the value received doesn't exceed 32 bytes. There's a test that fails if I don't normalize the value.

**Comment by voith** [comment]
@SamWilsn @lightclient This is ready for review.

File: src/ethereum/frontier/vm/instructions/memory.py
Comment (SamWilsn) [inline src/ethereum/frontier/vm/instructions/memory.py:50]: I'm still somewhat unsure how this is different from [`U256.to_be_bytes32`][0]. Could you maybe provide an example of how this snippet behaves differently? Should we change the implementation of [`U256.to_be_bytes32`][0]?

[0]: https://github.com/ethereum/eth1.0-specs/blob/3cac857d3556898adfd2ea988adfa6a794a0b367/src/ethereum/base_types.py#L548

File: src/ethereum/frontier/vm/gas.py
Comment (SamWilsn) [inline src/ethereum/frontier/vm/gas.py:75]: What does this return if you try to `mstore` a bit of memory which already exists? Something like:

```python
from ethereum.frontier.vm.gas import calculate_gas_for_extending_memory 

memory = bytearray(b"\x00" * 256)
calculate_gas_for_extending_memory(memory, 0, 32)
```

I suspect it tries to return a negative number?

File: src/ethereum/utils.py
Comment (SamWilsn) [inline src/ethereum/utils.py:50]:
```suggestion
        The same value if it's a perfect multiple of 32
```

File: src/ethereum/utils.py
Comment (SamWilsn) [inline src/ethereum/utils.py:51]: Maybe the one place a slightly more mathematical explanation is better :stuck_out_tongue_closed_eyes: 

```suggestion
        else it returns the smallest multiple of 32 that is greater than `value`.
```

**Review by SamWilsn — COMMENTED** [review]
Besides a few small docstring suggestions, the only thing I need clarification on is `calculate_gas_for_extending_memory`.

File: src/ethereum/frontier/vm/gas.py
Comment (voith) [reply src/ethereum/frontier/vm/gas.py:75]: @SamWilsn Nice catch Sam! I'll add a fix for this.

File: src/ethereum/utils.py
Comment (voith) [reply src/ethereum/utils.py:51]: Thanks for helping me fix the docs. Docs is the part I struggle with the most being a non-native English speaker.

File: src/ethereum/frontier/vm/instructions/memory.py
Comment (voith) [reply src/ethereum/frontier/vm/instructions/memory.py]: I was confused earlier because `U256.to_be_bytes32` would raise `OverflowError` for `value > U256_MAX_VALUE`. But now that I think of it that case is not possible as the stack is limited to U256 numbers only.  
Thanks for catching this. I have applied your suggestion.

File: src/ethereum/frontier/vm/gas.py
Comment (voith) [reply src/ethereum/frontier/vm/gas.py:75]: fixed.

**Comment by voith** [comment]
Thanks for the excellent review @SamWilsn. I have applied all your suggestions. 
@lightclient It would be nice to get your review too.

File: src/ethereum/frontier/vm/gas.py
Comment (SamWilsn) [inline src/ethereum/frontier/vm/gas.py:99]: Not sure if you'd agree, but maybe an easier way of implementing this would be something like:

```suggestion
    after_size = ceil32(start_position + size)
    
    if after_size <= before_size:
        return U256(0)
```

File: src/ethereum/frontier/vm/instructions/memory.py
Comment (SamWilsn) [inline src/ethereum/frontier/vm/instructions/memory.py:49]: Not saying that anything is incorrect here, but `U256(1)` jumps out as odd compared to the following two `U256(32)`. If this is correct, maybe add a short comment describing what you're doing?

Comment (voith) [reply src/ethereum/frontier/vm/instructions/memory.py:49]: maybe I should clarify in `calculate_gas_for_extending_memory` that the size parameter is the number of **bytes** by which we want to extend the memory. I can add comment here too.

Comment (SamWilsn) [reply src/ethereum/frontier/vm/instructions/memory.py:49]: So here we check the price for extending by one byte, and if it costs anything to add one byte, we change the cost to extend by 32 bytes?

Comment (voith) [reply src/ethereum/frontier/vm/instructions/memory.py:49]: Yeah, that's because the memory is a word addressable array. That logic is handled by the `ceil32` method. 
not sure If I understood your question completely and answered it.

File: src/ethereum/frontier/vm/gas.py
Comment (voith) [reply src/ethereum/frontier/vm/gas.py]: Please take a look at 850f028 and see if you'd be okay with that. I decided to club all of the logic into a single method and called it `extend_memory_and_subtract_gas`. I can change it if you don't like it.

**Comment by voith** [comment]
@SamWilsn I and @Bhargavasomu have mutually decided that we will merge this PR. I will him add as a co-author. Please let me know when you're happy with the code. I will squash the commits and add Somu as co-author before merging.

File: src/ethereum/frontier/vm/memory.py
Comment (Bhargavasomu) [inline src/ethereum/frontier/vm/memory.py:89]:
```suggestion
    for idx, byte in enumerate(value):
        memory[start_position + idx] = byte
```

File: src/ethereum/frontier/vm/memory.py
Comment (Bhargavasomu) [inline src/ethereum/frontier/vm/memory.py:68]: As far as the memory expansion goes in the yellow paper, I don't think the memory can ever shrink in a single transaction execution. Hence this condition can maybe be replaced with a comment mentioning that this condition isn't checked as it's assumed no entity will use it with a negative size.

\cc @SamWilsn

File: src/ethereum/frontier/vm/memory.py
Comment (Bhargavasomu) [inline src/ethereum/frontier/vm/memory.py:66]:
```suggestion
    current_memory_size = Uint(len(memory))
    expanded_memory_size = ceil32(start_position + size)
```
I feel these variables add more readability. Theoretically, the memory size is not limited and hence we should be using `Uint` instead of `U256`. Same goes for the output of ceil32 in above.

File: src/ethereum/frontier/vm/memory.py
Comment (Bhargavasomu) [inline src/ethereum/frontier/vm/memory.py:70]: image

Seems like doing the following is faster.
```
memory += b"\x00" * size_to_extend
```

File: src/ethereum/frontier/vm/memory.py
Comment (Bhargavasomu) [inline src/ethereum/frontier/vm/memory.py:64]: I think both these functions can be merged into a function called `expand_memory` or `extend_memory`. Since it's already mentioned in the docstring, I don't think it should be part of the function name.

FWIW, the condition `if size == 0: return None` should be present in the `extend_memory_and_subtract_gas` function.

I also prefer using `current_memory_size` and `expanded_memory_size` instead of `before_size` and `after_size` respectively.

Given that I don't think the memory can shrink, the following condition seems unnecessary.
```
if total_cost < already_paid:
    return None
```

File: src/ethereum/frontier/vm/memory.py
Comment (Bhargavasomu) [inline src/ethereum/frontier/vm/memory.py:21]: I also feel that we can get away with only 2 parameters to this function being the following:

- evm: Evm
- memory_size: U256

But this goes against how the yellowpaper defines the expansion. @SamWilsn @lightclient let me know your thoughts on this. Thanks.

**Review by Bhargavasomu — CHANGES_REQUESTED** [review]

File: src/ethereum/frontier/vm/memory.py
Comment (voith) [reply src/ethereum/frontier/vm/memory.py:21]: for the sake of consistency, I prefer to keep the pop statements as is i.e inside the method that implements the Opcode.

File: src/ethereum/frontier/vm/memory.py
Comment (voith) [reply src/ethereum/frontier/vm/memory.py:68]: I need to check this thoroughly. I'll compare the results with a working EVM.

File: tests/frontier/vm/test_memory_operations.py
Comment (voith) [reply tests/frontier/vm/test_memory_operations.py:32]: These tests raised an `OutOfGasError`. Upon investigation, I learned that the test is meant to raise an exception.
I also learned that fixtures that don't have a post-state are assumed to error out.

File: src/ethereum/frontier/vm/gas.py
Comment (SamWilsn) [reply src/ethereum/frontier/vm/gas.py:56]: LGTM

File: src/ethereum/frontier/vm/gas.py
Comment (SamWilsn) [reply src/ethereum/frontier/vm/gas.py:23]: I probably have an unpopular opinion here, but if a value is only used in one place and has no obvious meaningful name, putting it into a named constant just makes things harder to follow.

```python
quadratic_cost = size_in_words ** 2 // GAS_MEMORY_QUADRATIC_DENOMINATOR
```

vs.

```python
quadratic_cost = size_in_words ** 2 // 512
```

I think the second one is clearly easier to read.

That said, I know this is a contentious opinion, so I won't let it stand in the way of merging.

File: src/ethereum/frontier/vm/instructions/memory.py
Comment (SamWilsn) [reply src/ethereum/frontier/vm/instructions/memory.py:93]: `memeory` is my new favourite word.

File: src/ethereum/frontier/vm/memory.py
Comment (SamWilsn) [reply src/ethereum/frontier/vm/memory.py:64]: > Given that I don't think the memory can shrink, the following condition seems unnecessary.

It's true that the memory can't shrink, but you can initially `mload(100)` and later `mload(0)`. For the second `mload`, if I'm not mistaken, that would make `total_cost - already_paid` negative?

File: tests/frontier/vm/test_memory_operations.py
Comment (Bhargavasomu) [reply tests/frontier/vm/test_memory_operations.py:32]: Yes, that's the expected end state of those test files.

File: src/ethereum/frontier/vm/memory.py
Comment (Bhargavasomu) [reply src/ethereum/frontier/vm/memory.py:64]: Ahh, I now understand what's happening here and why those checks were enforced.

@voith I think you need to check if the memory which we are trying to access is already available or not. If it's already available the memory expansion shouldn't happen. Enforcing this check is much more clearer and readable than checking if the new gas is less than the already paid gas.

I leave it up to you whether you want to enforce this check in external code wherever this function is being called or inside this function itself. I weakly prefer having this check outside this function. You can refer to [this condition](https://github.com/ethereum/eth1.0-specs/pull/260/files#diff-5858d8f65876b560db7245474aad6c9acc689d7b6c887a008929cad952b32882R50-R53) in my old PR which actually implements this check.

File: src/ethereum/frontier/vm/memory.py
Comment (voith) [reply src/ethereum/frontier/vm/memory.py:112]: I just thought I'd match it with the signature of `U256.from_be_bytes`.

File: src/ethereum/frontier/vm/instructions/memory.py
Comment (voith) [reply src/ethereum/frontier/vm/instructions/memory.py:74]: I prefer to keep it as is. Memory extension happens before writing to memory, so If there is an OutofGasException I'd like to follow the trace sequentially.

File: src/ethereum/frontier/vm/instructions/memory.py
Comment (voith) [reply src/ethereum/frontier/vm/instructions/memory.py:47]: I prefer to keep it as is. Memory extension happens before writing to memory, so If there is an OutofGasException I'd like to follow the trace sequentially.

File: src/ethereum/frontier/vm/instructions/memory.py
Comment (Bhargavasomu) [reply src/ethereum/frontier/vm/instructions/memory.py:78]: Well, AFAIU I don't think it's economical to continue with memory expansion, given that you don't have enough Gas to even start the operation. It's a general pattern we follow in the rest of the opcode instructions as well. We first try to subtract the gas required for kicking off the operation, and later go to the internals of the operation.

\cc @SamWilsn @lightclient

File: src/ethereum/frontier/vm/instructions/memory.py
Comment (Bhargavasomu) [reply src/ethereum/frontier/vm/instructions/memory.py:49]: Well, AFAIU I don't think it's economical to continue with memory expansion, given that you don't have enough Gas to even start the operation. It's a general pattern we follow in the rest of the opcode instructions as well. We first try to subtract the gas required for kicking off the operation, and later go to the internals of the operation.

\cc @SamWilsn @lightclient

File: src/ethereum/frontier/vm/memory.py
Comment (Bhargavasomu) [reply src/ethereum/frontier/vm/memory.py]: I think `bytes` and `bytearray` are kind of synonymous as far as mypy is concerned. Hence I think it's ok if you drop the `bytes` wrapper.

File: src/ethereum/frontier/vm/instructions/memory.py
Comment (voith) [reply src/ethereum/frontier/vm/instructions/memory.py:49]: This is a spec, not a client. Readability triumphs over efficiency. For me reading the code in the order it occurs adds to better readability.

Comment (Bhargavasomu) [reply src/ethereum/frontier/vm/instructions/memory.py:49]: I never mentioned anything about efficiency. For me it doesn't make sense to even reach the memory expansion bit if you don't even have sufficient gas to power the opcode.

My philosophy here is that you can't run a car unless it has enough power to at least start.

Comment (voith) [reply src/ethereum/frontier/vm/instructions/memory.py:49]: My real reason for my approach is this section of the YP. 
Screenshot 2021-07-15 at 11 25 25 PM
for better readability, I decided to have the gas logic as consecutive statements so that its easy to see the actual cost function for memory operations and that is very closely related to how the YP defines it.  
Up until now all the opcodes implemented had `Cmemory = 0`. So it was easier to have the gas logic right at the start of the method.

> My philosophy here is that you can't run a car unless it has enough power to at least start.

I'm going to nitpick on some code that you wrote for the sake of this discussion(No malice intended just healthy debate):
https://github.com/ethereum/eth1.0-specs/blob/308b8ced06daeefca14d1b2c47ff0336e0d6cd5f/src/ethereum/frontier/vm/instructions/arithmetic.py#L305-L315
Going by your philosophy, why didn't you do?:
```python
evm.gas_left = subtract_gas(evm.gas_left, GAS_EXPONENTIATION)

if exponent != 0:
     exponent_bits = exponent.bit_length()
     exponent_bytes = (exponent_bits + 7) // 8
     evm.gas_left = subtract_gas(evm.gas_left, GAS_EXPONENTIATION * exponent_bytes)
```

Comment (Bhargavasomu) [reply src/ethereum/frontier/vm/instructions/memory.py:49]: @voith I don't think you understood my philosophy correctly. As my philosophy states, in the code that you highlighted, I am predicting the gas and deducting it previous to the actual opcode internals (which is to do the exponentiation).

But in your case, you are first calculating the cost needed for memory expansion and also expanding the memory which is an internal of the opcode, which is something none of our opcode implementations currently do (AFAIR). Let me know if I'm still not clear on my point.

I kindly request @SamWilsn @lightclient to opine on this issue. It's unclear how to proceed regarding this.

Comment (SamWilsn) [reply src/ethereum/frontier/vm/instructions/memory.py:49]: If I understand this debate correctly, the PR [as currently written](https://github.com/ethereum/eth1.0-specs/blob/ae4b619f3b5848cdbea0428beb01242283bd15b5/src/ethereum/frontier/vm/memory.py) does the following:

 1. Checks if there is enough gas for memory expansion.
 2. Expands the memory.
 3. Checks if there is enough gas for the non-variable portion of `mstore`.

The disagreement here is specifically about (2) coming before (3)?

The arguments for switching (2) and (3)—so that we do all the gas accounting up front and expand the memory after—are:
 * Sane order of operations (it's a waste to start the car if it'll run out of fuel as soon as you try to move); and
 * Consistency with how other opcodes are implemented.

That seems to make sense to me, but I might be misinterpreting @voith's position.

In the grand scheme of things, I don't see this making a huge difference either way, and even if it does, I think it's something we can fix in the future.

This discussion is blocking quite a bit of code, so whatever gets us unblocked the quickest is my personal preference (and _please_ don't consider me as an ultimate arbiter for resolving PRs; I don't want that responsibility!)

Comment (voith) [reply src/ethereum/frontier/vm/instructions/memory.py]: @SamWilsn @Bhargavasomu Take a look at d5b38d6d647cc6302a7e00889f9c6a6c4bb3cdca
I split the gas calculation method into a separate method. It's a bit more verbose but it exactly follows the yellow paper. 
Let me know what you think.

File: src/ethereum/frontier/vm/memory.py
Comment (voith) [reply src/ethereum/frontier/vm/memory.py]: I have used `before_size` and `after_size` which IMO are quite similar to this suggestion here. However, @SamWilsn I'll commit this if you feel strongly about it.

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**
